aws/protocol/rest/encode.go

```diff
@@ -0,0 +1,79 @@
+package rest
+
+import (
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+// An Encoder provides encoding of REST URI path, query, and header components
+// of an HTTP request. Can also encode a stream as the payload.
+//
+// Does not support SetFields.
+type Encoder struct {
+	req *http.Request
+
+	path, rawPath, pathBuffer []byte
+
+	query  url.Values
+	header http.Header
+}
+
+// NewEncoder creates a new encoder from the passed in request. All query and
+// header values will be added on top of the request's existing values. Overwriting
+// duplicate values.
+func NewEncoder(req *http.Request) *Encoder {
+	e := &Encoder{
+		req: req,
+
+		path:    []byte(req.URL.Path),
+		rawPath: []byte(req.URL.Path),
+		query:   req.URL.Query(),
+		header:  req.Header,
+	}
+
+	return e
+}
+
+// Encode returns a REST protocol encoder for encoding HTTP bindings
+// Returns any error if one occurred during encoding.
+func (e *Encoder) Encode() error {
+	e.req.URL.Path, e.req.URL.RawPath = string(e.path), string(e.rawPath)
+	e.req.URL.RawQuery = e.query.Encode()
+	e.req.Header = e.header
+
+	return nil
+}
```

Comment:
How will this return an error, if it occurred during encoding?

Reply:
Ok to return error in this case since, we may want to add validation in the future